leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/SecurityTest.java

```diff
@@ -131,6 +132,77 @@ public void registered_device_with_oscore_to_server_with_oscore()
         assertTrue(response.isSuccess());
     }
 
+    // TODO OSCORE should failed but does not because context by URI is not removed.
+    @Ignore
+    @Test
+    public void registered_device_with_oscore_to_server_with_oscore_then_removed_security_info_then_server_fails_to_send_request()
+            throws NonUniqueSecurityInfoException, InterruptedException {
+
+        helper.createOscoreServer();
+        helper.server.start();
+
+        helper.createOscoreClient();
+        helper.getSecurityStore()
+                .add(SecurityInfo.newOscoreInfo(helper.getCurrentEndpoint(), getServerOscoreSetting()));
+
+        // Check client is not registered
+        helper.assertClientNotRegisterered();
+
+        // Start it and wait for registration
+        helper.client.start();
+        helper.waitForRegistrationAtServerSide(1);
+
+        // Check client is well registered
+        helper.assertClientRegisterered();
+
+        // check we can send request to client.
+        ReadResponse response = helper.server.send(helper.getCurrentRegistration(), new ReadRequest(3, 0, 1), 500);
+        assertTrue(response.isSuccess());
+
+        // Remove securityInfo
+        helper.getSecurityStore().remove(helper.getCurrentEndpoint(), true);
+
+        // check we can send request to client.
+        response = helper.server.send(helper.getCurrentRegistration(), new ReadRequest(3, 0, 1), 500);
+        assertTrue(!response.isSuccess());
+        // TODO OSCORE we must defined the expected behavior here.
+    }
```

Comment:
With this test, I am able to confirm that context from uriMap is not cleaned.

I tried to create my own `HashMapCtxDB` by copy/paste it as a shortterm workaround but  I can't because of : 
```java
@Override
	public synchronized void addContext(String uri, OSCoreCtx ctx) throws OSException {
		if (uri != null) {
			String normalizedUri = normalizeServerUri(uri);
			uriMap.put(normalizedUri, ctx);
			ctx.setUri(normalizedUri);
		}
		addContext(ctx);
	}
```
I can not call `ctx.setUri(normalizedUri);` because this is a protected method :grimacing: 

Any other idea for a shortterm workaround ? 

I guess if we want user being able to create their own `OSCoreCtxDB`, they will need to call `OSCoreCtx.setUri(String)`  or maybe this method should not be called here ?

Reply:
Yeah probably a number of methods should be made accessible.

As I recall the OSCoreCtx.setUri(String) method was originally needed to have the URI associated with a context accessible in the Appendix B.2. context rederivation procedure. I think only _OSCoreEndpointContextInfo_ and _ContextRederivation_ use the corresponding OSCoreCtx.getUri() method.